### PR TITLE
feat(module): Add a new module "secret" for managing a single secret

### PR DESCRIPTION
**Pull request by legal90**
This Pull Request adds a new module `./modules/secret`, which allows to create individual secrets and their versions.
The module is self-contained and could be used directly:

```hcl
module "secret-manager" {
  source  = "<path-to>//modules/secret"

  project_id  = var.project_id
  name        = "secret-1"
  secret_data = "secret information"
}
```

The purpose is to have a simple module with a straightforward and well-documented interface (variables).

Here I also updated the examples `kms`, `pubsub` and `simple` to demonstrate the simplicity of managing a single secret using a new module.

____________________
**To maintainers:** 
This PR doesn't affect the root module and is safe to merge. But if you are interested, I can submit another PR which updates the root module to use this new `./modules/secret` (wrapping it with `for_each`). That will remove a code duplications between these two modules.  You can see how it might look in this branch: [link](https://github.com/Billhop/terraform-google-secret-manager/commit/1a7264dfc9655edbc216b63108f780d987dc0e75)
But that introduces a breaking change for all existing users of a root module, that's why I didn't include it here.

### CODE REVIEW

**Comment by apeabody** [comment]
From the test:

```
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; 
        	            	Error: Output "project_number" not found
        	            	
        	            	The output variable requested could not be found in the state file. If you
        	            	recently added this to your configuration, be sure to run `terraform apply`,
        	            	since the state won't be updated with new output variables until that command
        	            	is run.}
        	Test:       	TestSimpleSecret
```

**Comment by legal90** [comment]
@apeabody Oh, I didn't notice that "simple" example module was used in the integration tests.
I updated it now. Could you please re-run the test?

**Comment by apeabody** [comment]
```
# github.com/terraform-google-modules/terraform-google-secretmanager/test/integration/simple [github.com/terraform-google-modules/terraform-google-secretmanager/test/integration/simple.test]
simple/simple_test.go:32:3: projectDescribe declared and not used
simple/simple_test.go:33:17: undefined: proj
FAIL	github.com/terraform-google-modules/terraform-google-secretmanager/test/integration/simple [build failed]
```

**Comment by apeabody** [comment]
Thanks @legal90 - My current thought is it would make sense to merge this new module/PR first, and then after the next release entertain a breaking PR to migrate the main module.  The PR does somewhat reduce the test coverage of the main module, but the later PR would basically restore that.

@bharathkkb WDYT?

File: modules/secret/README.md [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+## Requirements
+
+These sections describe requirements for using this module.
+
+### Software
+
+The following dependencies must be available:
+
+- [Terraform][terraform] v0.13
+- [Terraform Provider for GCP][terraform-provider-gcp] plugin v3.0
```
Comment (apeabody) [inline modules/secret/README.md:57]:
```suggestion
- [Terraform Provider for GCP][terraform-provider-gcp] plugin >=v4.83.0
```

File: modules/secret/variables.tf [hunk 2]
```diff
@@ -0,0 +1,91 @@
+/**
+ * Copyright 2022 Google LLC
```
Comment (apeabody) [inline modules/secret/variables.tf:2]: ?

```suggestion
 * Copyright 2024 Google LLC
```

File: modules/secret/outputs.tf [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/**
+ * Copyright 2022 Google LLC
```
Comment (apeabody) [inline modules/secret/outputs.tf:2]: ?

```suggestion
 * Copyright 2024 Google LLC
```

File: modules/secret/main.tf [hunk 4]
```diff
@@ -0,0 +1,73 @@
+/**
+ * Copyright 2022 Google LLC
```
Comment (apeabody) [inline modules/secret/main.tf:2]: ?

```suggestion
 * Copyright 2024 Google LLC
```

File: modules/secret/versions.tf [hunk 5]
```diff
@@ -0,0 +1,33 @@
+/**
+ * Copyright 2021 Google LLC
```
Comment (apeabody) [inline modules/secret/versions.tf:2]: ?

```suggestion
 * Copyright 2024 Google LLC
```

File: modules/secret/versions.tf [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      version = ">= 4.83.0, < 6"
+    }
+    google-beta = {
+      source  = "hashicorp/google-beta"
+      version = ">= 4.83.0, < 6"
+    }
+  }
+
+  provider_meta "google" {
+    module_name = "blueprints/terraform/secret-manager/v0.2.0"
```
Comment (apeabody) [inline modules/secret/versions.tf:31]:
```suggestion
    module_name = "blueprints/terraform/secret-manager/v0.3.0"
```

File: modules/secret/README.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
+# terraform-google-secret-manager
```
Comment (apeabody) [inline modules/secret/README.md:1]:
```suggestion
# Terraform Google Secret Manager Secret Submodule
```

File: modules/secret/README.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+This modules makes it easy to create a single Google Secret Manager secret. It supports using custom KMS keys (CMEK) for encrypting the secrets.
```
Comment (apeabody) [inline modules/secret/README.md:3]:
```suggestion
This module makes it easy to create a single Google Secret Manager secret. It supports using custom KMS keys (CMEK) for encrypting the secrets.
```

File: modules/secret/README.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  source  = "GoogleCloudPlatform/secret-manager/google//modules/secret"
+  version = "~> 0.3"
+
+  project_id  = var.project_id
+  name        = "secret-1"
+  secret_data = "secret information"
+}
+```
+
+Functional examples are included in the [examples](./examples/) directory.
```
Comment (apeabody) [inline modules/secret/README.md:21]: Let's call out the specific example which uses this particular submodule.

**Review by apeabody — CHANGES_REQUESTED** [review]
Thanks for the contribution @legal90!

File: modules/secret/README.md (hunk 9 above)
Comment (legal90) [reply modules/secret/README.md]: Done in https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/pull/76/commits/fc61bc59118354620bb39c52f73a4384c6dc725e

**Comment by apeabody** [comment]
@bharathkkb WDYT?

File: modules/secret/README.md [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+This module makes it easy to create a single Google Secret Manager secret. It supports using custom KMS keys (CMEK) for encrypting the secrets.
+It also supports secret rotation and can configure the rotation notification to be sent to given pubsub topics.
+
+## Usage
+
+Basic usage of this module is as follows:
+
+```hcl
+module "secret-manager" {
+  source  = "GoogleCloudPlatform/secret-manager/google//modules/secret"
```
Comment (bharathkkb) [inline modules/secret/README.md:12]: For consistency with some of our other single variants let's call it simple-secret here and throughout.
```suggestion
  source  = "GoogleCloudPlatform/secret-manager/google//modules/simple-secret"
```

File: examples/pubsub/main.tf [hunk 11]
```diff
@@ -36,26 +36,19 @@ resource "google_pubsub_topic_iam_member" "sm_sa_publisher" {
   topic   = google_pubsub_topic.secret.name
 }
 
-module "secret-manager" {
-  source  = "GoogleCloudPlatform/secret-manager/google"
+module "secret" {
+  source  = "GoogleCloudPlatform/secret-manager/google//modules/secret"
```
Comment (bharathkkb) [inline examples/pubsub/main.tf:40]: Let's change the simple example but keep other showcasing root module

File: modules/secret/versions.tf [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+      version = ">= 4.83.0, < 6"
+    }
+    google-beta = {
+      source  = "hashicorp/google-beta"
+      version = ">= 4.83.0, < 6"
+    }
+  }
+
+  provider_meta "google" {
+    module_name = "blueprints/terraform/secret-manager/v0.3.0"
```
Comment (bharathkkb) [inline modules/secret/versions.tf:31]:
```suggestion
    module_name = "blueprints/terraform/secret-manager:simple-secret/v0.3.0"
```

**Review by bharathkkb — COMMENTED** [review]
A few nits but LGTM overall. The main benefit I see with this module is for cases where name is dynamically derived so I think wrapping with main module would run into for_each cannot determine resource count errors, so prefer not to do that.

File: modules/secret/README.md (hunk 10 above)
Comment (legal90) [reply modules/secret/README.md]: OK, renamed it to `simple-secret` 👍 .

File: examples/pubsub/main.tf (hunk 11 above)
Comment (legal90) [reply examples/pubsub/main.tf]: OK, done.

**Comment by legal90** [comment]
Thank you for the review, @bharathkkb ! 

P.s.
Actually, wrapping the module into `for_each` works fine (as shown here: https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/commit/a11dcee366f276a569cf5f0fbc17696e4d34f6a4#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR48-R65).  `tarraform plan` in `examples/multiple` from that branch compiles successfully.

Anyway, I applied the changes which you suggested. Is it good now?

**Comment by apeabody** [comment]
> Thank you for the review, @bharathkkb !
> 
> Actually, wrapping the module into `for_each` works fine (as shown here: [a11dcee#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR48-R65](https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/commit/a11dcee366f276a569cf5f0fbc17696e4d34f6a4#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR48-R65)). `tarraform plan` in `examples/multiple` from that branch compiles successfully.
> 
> Anyway, I applied the changes which you suggested. Is it good now?

Thanks @legal90 @bharathkkb! - I plan to do a final review today.

File: modules/simple-secret/README.md [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+It also supports secret rotation and can configure the rotation notification to be sent to given pubsub topics.
+
+## Usage
+
+Basic usage of this module is as follows:
+
+```hcl
+module "secret-manager" {
+  source  = "GoogleCloudPlatform/secret-manager/google//modules/simple-secret"
+  version = "~> 0.3"
```
Comment (apeabody) [inline modules/simple-secret/README.md:13]: sync to the recent release

```suggestion
  version = "~> 0.4"
```

Comment (apeabody) [reply modules/simple-secret/README.md:13]: Should be ready to merge after this update.

**Comment by bharathkkb** [comment]
> Actually, wrapping the module into `for_each` works fine (as shown here: [a11dcee#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR48-R65](https://github.com/GoogleCloudPlatform/terraform-google-secret-manager/commit/a11dcee366f276a569cf5f0fbc17696e4d34f6a4#diff-dc46acf24afd63ef8c556b77c126ccc6e578bc87e3aa09a931f33d9bf2532fbbR48-R65)). `tarraform plan` in `examples/multiple` from that branch compiles successfully.

Ah reverse should be fine! I misread the proposal as we could create a single instance of the root module within this new simple module to create a single secret. Still let's defer this to a future PR due to the migration concern you pointed out.

**Review by apeabody — APPROVED** [review]

**Pull request merged**